### PR TITLE
gpu monitoring

### PR DESCRIPTION
**Pull request by vishiy**

### CODE REVIEW

File: source/code/plugin/in_kube_nodes.rb [hunk 1]
```diff
@@ -191,6 +193,20 @@ def parse_and_emit_records(nodeInventory, batchTime = Time.utc.iso8601)
             capacityInfo = items["status"]["capacity"]
             ApplicationInsightsUtility.sendMetricTelemetry("NodeMemory", capacityInfo["memory"], properties)
 
+            begin
+              if (!capacityInfo["nvidia.com/gpu"].nil?) && (!capacityInfo["nvidia.com/gpu"].empty?)
+                properties["nvigpus"] = capacityInfo["nvidia.com/gpu"]
+              end
+
+              if (!capacityInfo["amd.com/gpu"].nil?) && (!capacityInfo["amd.com/gpu"].empty?)
```
Comment (joby-thomas) [inline source/code/plugin/in_kube_nodes.rb]: Is it a supported scenario to have both nvidia & AMD gpu in a single VM? if not its better to move the amd check as else if ?

File: source/code/plugin/CAdvisorMetricsAPIClient.rb [hunk 2]
```diff
@@ -282,6 +283,101 @@ def getContainerCpuMetricItems(metricJSON, hostName, cpuMetricNameToCollect, met
...
+          metricInfo = JSON.parse(cAdvisorStats.body)
+        end
+        if !winNode.nil?
+          hostName = winNode["Hostname"]
+          operatingSystem = "Windows"
+        else
+          if !metricInfo.nil? && !metricInfo["node"].nil? && !metricInfo["node"]["nodeName"].nil?
+            hostName = metricInfo["node"]["nodeName"]
+          else
+            hostName = (OMS::Common.get_hostname)
```
Comment (joby-thomas) [inline source/code/plugin/CAdvisorMetricsAPIClient.rb]: Is this code is assuming it is running in the Windows Node?

File: source/code/plugin/CAdvisorMetricsAPIClient.rb [hunk 3]
```diff
@@ -282,6 +283,101 @@ def getContainerCpuMetricItems(metricJSON, hostName, cpuMetricNameToCollect, met
...
+                    metricItems.push(metricItem)
+                  end
+                end
+              end
+            end
+          end
+        end
+      rescue => errorStr
+        @Log.warn("getContainerGpuMetricsAsInsightsMetrics failed: #{errorStr} for metric #{metricNameToCollect}")
+        return metricItems
```
Comment (joby-thomas) [inline source/code/plugin/CAdvisorMetricsAPIClient.rb]: if the next line is return metricItems do we need this inside the rescue? Ruby wont execute the next line?

File: source/code/plugin/CAdvisorMetricsAPIClient.rb (hunk 2 above)
Comment (vishiy) [reply source/code/plugin/CAdvisorMetricsAPIClient.rb]: we dont have a presence in windows nodes. This is windows remote.

File: source/code/plugin/CAdvisorMetricsAPIClient.rb (hunk 3 above)
Comment (vishiy) [reply source/code/plugin/CAdvisorMetricsAPIClient.rb]: return ends the method.

File: source/code/plugin/in_kube_nodes.rb (hunk 1 above)
Comment (vishiy) [reply source/code/plugin/in_kube_nodes.rb]: This is specifically meant for hybrids. EKS,GKE & AKS dont have mixed gpu SKUs (yet), but hybrids can have both in the same node, as they are just accelerator cards that goes into the ePCI slots.

File: source/code/plugin/in_cadvisor_perf.rb [hunk 4]
```diff
@@ -68,6 +70,31 @@ def enumerate()
...
+            wrapper = {
+              "DataType" => "INSIGHTS_METRICS_BLOB",
+              "IPName" => "ContainerInsights",
+              "DataItems" => [insightsMetricsRecord.each { |k, v| insightsMetricsRecord[k] = v }],
+            }
+            insightsMetricsEventStream.add(time, wrapper) if wrapper
+          end
+
+          router.emit_stream(Constants::INSIGHTSMETRICS_FLUENT_TAG, insightsMetricsEventStream) if insightsMetricsEventStream
+          $log.info("cAdvisorInsightsMetricsEmitStreamSuccess @ #{Time.now.utc.iso8601}")
```
Comment (rashmichandrashekar) [inline source/code/plugin/in_cadvisor_perf.rb]: pls check for test env variable -
@@istestvar = ENV["ISTEST"]
        if (!@@istestvar.nil? && !@@istestvar.empty? && @@istestvar.casecmp("true") == 0 && eventStream.count > 0)
          $log.info("kubePodInventoryEmitStreamSuccess @ #{Time.now.utc.iso8601}")
        end

File: source/code/plugin/in_kube_nodes.rb [hunk 5]
```diff
@@ -235,6 +251,32 @@ def parse_and_emit_records(nodeInventory, batchTime = Time.utc.iso8601)
...
+
+            nodeGPUInsightsMetricsDataItems.each do |insightsMetricsRecord|
+              wrapper = {
+                "DataType" => "INSIGHTS_METRICS_BLOB",
+                "IPName" => "ContainerInsights",
+                "DataItems" => [insightsMetricsRecord.each { |k, v| insightsMetricsRecord[k] = v }],
+              }
+              insightsMetricsEventStream.add(emitTime, wrapper) if wrapper
+            end
+
```
Comment (rashmichandrashekar) [inline source/code/plugin/in_kube_nodes.rb]: pls add success log line for test scenario

File: source/code/plugin/in_kube_podinventory.rb [hunk 6]
```diff
@@ -496,6 +498,33 @@ def parse_and_emit_records(podInventory, serviceList, batchTime = Time.utc.iso86
...
+            containerGPUInsightsMetricsDataItems.each do |insightsMetricsRecord|
+              wrapper = {
+                "DataType" => "INSIGHTS_METRICS_BLOB",
+                "IPName" => "ContainerInsights",
+                "DataItems" => [insightsMetricsRecord.each { |k, v| insightsMetricsRecord[k] = v }],
+              }
+              insightsMetricsEventStream.add(emitTime, wrapper) if wrapper
+            end
+
+            router.emit_stream(Constants::INSIGHTSMETRICS_FLUENT_TAG, insightsMetricsEventStream) if insightsMetricsEventStream
```
Comment (rashmichandrashekar) [inline source/code/plugin/in_kube_podinventory.rb]: pls add test log line after checking the env var

File: source/code/plugin/in_win_cadvisor_perf.rb [hunk 7]
```diff
@@ -82,6 +84,30 @@ def enumerate()
...
+              wrapper = {
+                "DataType" => "INSIGHTS_METRICS_BLOB",
+                "IPName" => "ContainerInsights",
+                "DataItems" => [insightsMetricsRecord.each { |k, v| insightsMetricsRecord[k] = v }],
+              }
+              insightsMetricsEventStream.add(time, wrapper) if wrapper
+            end
+
+            router.emit_stream(Constants::INSIGHTSMETRICS_FLUENT_TAG, insightsMetricsEventStream) if insightsMetricsEventStream
+            $log.info("winCAdvisorInsightsMetricsEmitStreamSuccess @ #{Time.now.utc.iso8601}")
```
Comment (rashmichandrashekar) [inline source/code/plugin/in_win_cadvisor_perf.rb]: pls add env variable check

File: source/code/plugin/in_kube_nodes.rb (hunk 5 above)
Comment (vishiy) [reply source/code/plugin/in_kube_nodes.rb]: fixed

File: source/code/plugin/in_kube_podinventory.rb (hunk 6 above)
Comment (vishiy) [reply source/code/plugin/in_kube_podinventory.rb]: fixed

**Pull request merged**